File: Medium/725. Split Linked List in Parts/725. Split Linked List in Parts.ts
/**
 * Definition for singly-linked list.
 * class ListNode {
 *     val: number
 *     next: ListNode | null
 *     constructor(val?: number, next?: ListNode | null) {
 *         this.val = (val === undefined ? 0 : val);
 *         this.next = (next === undefined ? null : next);
 *     }
 * }
 */

function splitListToParts(
    head: ListNode | null,
    k: number
): Array<ListNode | null> {
    // Calculate the length of the linked list
    let length = 0;
    let current: ListNode | null = head;
    while (current !== null) {
        length++;
        current = current.next;
    }

    // Calculate the size of each part and the number of extra nodes
    const partSize = Math.floor(length / k);
    const extraNodes = length % k;

    const result: Array<ListNode | null> = [];
    current = head;

    for (let i = 0; i < k; i++) {
        result.push(current);
        const partLength = partSize + (i < extraNodes ? 1 : 0);

        // Move current to the end of this part
        for (let j = 1; j < partLength && current !== null; j++) {
            current = current.next;
        }

        if (current !== null) {
            // Save the next node, break the link, and move to the next part
            const nextNode: ListNode | null = current.next;
            current.next = null;
            current = nextNode;
        }
    }

    return result;
}

File: Medium/725. Split Linked List in Parts/725. Split Linked List in Parts.rs
// Definition for singly-linked list.
// #[derive(PartialEq, Eq, Clone, Debug)]
// pub struct ListNode {
//   pub val: i32,
//   pub next: Option<Box<ListNode>>
// }
// 
// impl ListNode {
//   #[inline]
//   fn new(val: i32) -> Self {
//     ListNode {
//       next: None,
//       val
//     }
//   }
// }
impl Solution {
    pub fn split_list_to_parts(head: Option<Box<ListNode>>, k: i32) -> Vec<Option<Box<ListNode>>> {
        // Count the length of the linked list
        let mut current = &head;
        let mut length = 0;
        while let Some(node) = current {
            length += 1;
            current = &node.next;
        }
        
        // Calculate the size of each part and the number of larger parts
        let part_size = length / k;
        let mut larger_parts = length % k; // Declare as mutable
        
        let mut result = Vec::new();
        let mut current = head;
        
        for _ in 0..k {
            let mut part_head = None;
            let mut part_tail = &mut part_head;
            
            for _ in 0..part_size + if larger_parts > 0 { 1 } else { 0 } {
                if let Some(mut node) = current.take() {
                    current = node.next.take();
                    *part_tail = Some(node);
                    part_tail = &mut part_tail.as_mut().unwrap().next;
                }
            }
            
            result.push(part_head);
            
            if larger_parts > 0 {
                larger_parts -= 1;
            }
        }
        
        result
    }
}